### ISSUE TITLE
Role Assignment Upon Successful Purchase

### ISSUE THREAD

**Issue by jeffolmstead** [OrchardCMS/OrchardCore.Commerce#7]
The goal would be to have a "Part" (i.e. RoleAssignmentPart) that you would add to a Product where the user would toggle which Role to assign to the purchaser at time of successful purchase.

At the time of successful purchase, this role would automatically be added to the user. As an alternative, it could be handled by a workflow but that would require the user to manually handle this themselves (or use a recipe go generate when feature is enabled).

As a related example of usage, I have used a similar concept to put the user into a role which gives them access to a page / download for 30 days. I have a secondary workflow (not related to this module) which removes the role automatically.

**Comment by bleroy** [comment]
It seems like a role assignment activity in workflows could be contributed to Orchard Core, as its not specific to commerce. The commerce part of this proposal seems to be a purchase workflow activity.

**Comment by jeffolmstead** [comment]
@bleroy if I understand your comment correctly, it means if a user wants a role assignment they have to:

1. Add in the product (nothing unique on product about roles)
2. Navigate and edit a workflow (likely containing a product selector) to have this product assign the role once purchased

Personally, I don't mind this separation of concerns but in some organizations a users role might not let them edit the workflow. This is why I made it as a part so it could be controlled right on the product. It does mean more code to maintain though (and I prefer native when possible) so I can be sold on @bleroy approach.

**Comment by bleroy** [comment]
Yes. We could have recipes that set-up common or basic workflows though. It gives you a starting point, makes it easier to discover, and still allows you to extend and modify later.

**Comment by jeffolmstead** [comment]
I am a huge fan of Workflows so this would be my recommendation (and possibly we don't need to implement this as much as keep it a consideration during buildout). 

Perhaps this is a bit more OrchardCore of a comment, but what would be ideal is if you were on a Product (or any Content) and there was enough logic to pull in a section of "Workflows Related To this Item". That would provide a ton of convenience and really make Workflows a core "hub" of the content system as you could quickly drill in. Any feasibility to this? I could port this over to OrchardCore as an enhancement / discussion.

**Comment by bleroy** [comment]
Absolutely. One of the ways Orchard could be improved in my opinion is in connecting various components that are logically related, but that are all over the place in Orchard. In other words, implementation details are a little too visible. I wish you could, as you're suggesting, integrate workflows with items, but more generally, have admin tasks built on top of Orchard that enable the implementation by the integrator of business processes in a way that makes sense to the end user, without them having to learn what a part is, or a workflow, or media, and so on.

**Comment by jeffolmstead** [comment]
Carried that over to OrchardCore issues [here](https://github.com/OrchardCMS/OrchardCore/issues/2578).
